File: apps/mobile/android/app/src/main/java/com/visvine/mobile/data/repository/FeedRepository.kt
package com.visvine.mobile.data.repository

import com.visvine.mobile.data.model.FeedPage
import com.visvine.mobile.data.remote.ApiResult
import com.visvine.mobile.data.remote.MediaUrl
import com.visvine.mobile.data.remote.VisvineApi
import com.visvine.mobile.data.remote.safeApiCall
import kotlinx.serialization.json.Json
import javax.inject.Inject
import javax.inject.Singleton

/** The space's feed — posts in the FEED-mode channels the caller has joined. */
@Singleton
class FeedRepository @Inject constructor(
    private val api: VisvineApi,
    private val json: Json,
) {
    suspend fun getFeed(spaceId: String?, cursor: String? = null): ApiResult<FeedPage> =
        when (val res = safeApiCall(json) { api.getFeed(spaceId, cursor) }) {
            is ApiResult.Success -> ApiResult.Success(
                res.data.copy(
                    posts = res.data.posts.map { post ->
                        post.copy(
                            message = post.message.copy(
                                sender = post.message.sender.copy(image = MediaUrl.resolve(post.message.sender.image)),
                                attachmentUrl = MediaUrl.resolve(post.message.attachmentUrl),
                            ),
                        )
                    },
                ),
            )
            is ApiResult.Failure -> res
        }
}

File: apps/mobile/ios/Visvine/Repositories/FeedRepository.swift
import Foundation

/// One space's feed — GET /api/feed?spaceId=&cursor= (docs/mobile.md § Home).
struct FeedRepository {
    private let api = APIClient.shared

    func getFeed(spaceId: String, cursor: String? = nil) async -> APIResult<FeedPage> {
        let res: APIResult<FeedPage> = await api.request("/api/feed", query: ["spaceId": spaceId, "cursor": cursor, "limit": "20"])
        switch res {
        case .success(let page):
            var copy = page
            copy.posts = page.posts.map { post in
                var p = post
                p.message.sender.image = MediaURL.resolve(post.message.sender.image)
                return p
            }
            return .success(copy)
        case .failure(let m):
            return .failure(m)
        }
    }
}
